shell/common/platform_util_linux.cc

```diff
@@ -32,54 +49,61 @@ bool XDGUtil(const std::vector<std::string>& argv, const bool wait_for_exit) {
   if (!process.IsValid())
     return false;
 
-  if (wait_for_exit) {
+  if (wait_for_exit || !callback.is_null()) {
     int exit_code = -1;
     if (!process.WaitForExit(&exit_code))
       return false;
+
+    if (!callback.is_null())
+      std::move(callback).Run(GetErrorDescription(exit_code));
+
     return (exit_code == 0);
   }
 
   base::EnsureProcessGetsReaped(std::move(process));
   return true;
 }
 
-bool XDGOpen(const std::string& path, const bool wait_for_exit) {
-  return XDGUtil({"xdg-open", path}, wait_for_exit);
+bool XDGOpen(const std::string& path,
+             const bool wait_for_exit,
+             platform_util::OpenCallback callback) {
+  return XDGUtil({"xdg-open", path}, wait_for_exit, std::move(callback));
 }
 
 bool XDGEmail(const std::string& email, const bool wait_for_exit) {
-  return XDGUtil({"xdg-email", email}, wait_for_exit);
+  return XDGUtil({"xdg-email", email}, wait_for_exit,
+                 platform_util::OpenCallback());
 }
 
 }  // namespace
 
 namespace platform_util {
 
-// TODO(estade): It would be nice to be able to select the file in the file
-// manager, but that probably requires extending xdg-open. For now just
-// show the folder.
 void ShowItemInFolder(const base::FilePath& full_path) {
   base::FilePath dir = full_path.DirName();
   if (!base::DirectoryExists(dir))
     return;
 
-  XDGOpen(dir.value(), false);
+  XDGOpen(dir.value(), false, platform_util::OpenCallback());
 }
 
-bool OpenItem(const base::FilePath& full_path) {
-  return XDGOpen(full_path.value(), false);
+void OpenPath(const base::FilePath& full_path, OpenCallback callback) {
+  // This is async, so we don't care about the return value.
+  XDGOpen(full_path.value(), false, std::move(callback));
```

Comment:
Isn't the purpose of async version to extract error message from process exit value, why don't we care about exit value then ?

Reply:
this may have been unclear - by this i mean that we don't care about what `XDGOpen` returns; we're still calling back the exit code with that call.

Reply:
`wait_for_exit` should be set to `true` since we depend on the `exit_code`, and simplify the above check

```
if (wait_for_exit) {
  ...
  bool success = process.WaitForExit(&exit_code);
  if (!callback.is_null())
   std::move(callback).Run(..);
  return success ? (exit_code == 0) : false;
```